### PR TITLE
[misc][Long Context] feat: support ulysses for long context training

### PR DESCRIPTION
**Pull request by PeterSH6**
- To support ulysses, we implemented a `FSDPUlyssesShardingManager` to manage the SP Parallel states of different models. And we utilize device mesh to manage the SP parallel groups.
- The long context training is supported through monkey_patch. Currently, we support llama and qwen2 architectures. Will support other models later
- Before model fwd, we pad the input_ids to be divisible by SP size and then slice the input_ids. The position_ids are only padded not sliced to make sure the position_embedding can match the qkv_states. This can be optimized.
- Make shuffle to False in mini_batch_iterator

### CODE REVIEW

File: verl/models/transformers/__init__.py
Comment (vermouth1992) [inline verl/models/transformers/__init__.py:1]: Add license

File: tests/model/test_transformers_ulysses.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
+import torch
```
Comment (vermouth1992) [inline tests/model/test_transformers_ulysses.py:15]: Add license

Comment (vermouth1992) [reply tests/model/test_transformers_ulysses.py:15]: Do we have a CI to check license?

File: verl/models/transformers/monkey_patch.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+"""
+Apply monkey-patch function to models
+"""
+
+#### Open Source Models
+#### transformers version < 4.48
+
+
+def apply_monkey_patch_to_llama():
+    from transformers.models.llama.modeling_llama import LlamaFlashAttention2
```
Comment (vermouth1992) [inline verl/models/transformers/monkey_patch.py:23]: Shall we add an assertion of transformer version here?

File: verl/models/transformers/monkey_patch.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+_PATCH_NAME_TO_FUNC = {
+    'llama': apply_monkey_patch_to_llama,
+    'qwen2': apply_monkey_patch_to_qwen2,
+}
+
+from transformers import PretrainedConfig
+
+
+def apply_monkey_patch(config: PretrainedConfig, verbose=True):
+    success_apply_monkey_patch = False
```
Comment (vermouth1992) [inline verl/models/transformers/monkey_patch.py:46]: Shall we add an assertion of transformer version here?

File: verl/utils/ulysses.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
+from typing import Any, Optional, List, Tuple
```
Comment (vermouth1992) [inline verl/utils/ulysses.py:19]: Add license here. We should also acknowledge deepspeed and their paper

File: tests/model/test_transformers_ulysses.py (hunk 1 above)
Comment (PeterSH6) [reply tests/model/test_transformers_ulysses.py:15]: Not yet, I will establish one tomorrow.

Comment (vermouth1992) [reply tests/model/test_transformers_ulysses.py:15]: The license CI is up. Please rebase main

File: verl/workers/actor/dp_actor.py [hunk 5]
```diff
@@ -62,14 +64,29 @@ def _forward_micro_batch(self, micro_batch, temperature) -> Tuple[torch.Tensor,
...
                 # only pass input_ids and position_ids to enable flash_attn_varlen
                 output = self.actor_module(input_ids=input_ids_rmpad,
                                            attention_mask=None,
                                            position_ids=position_ids_rmpad,
                                            use_cache=False)  # prevent model thinks we are generating
                 logits_rmpad = output.logits.squeeze(0)  # (total_nnz, vocab_size)
+
+                # gather output if sp > 1
+                if self.ulysses_sequence_parallel_size > 1:
+                    logits_rmpad = gather_outpus_and_unpad(logits_rmpad,
```
Comment (vermouth1992) [inline verl/workers/actor/dp_actor.py]: It's problematic to perform any operation on full logits because it is very large. We should perform reduction on logits first, then gather

File: verl/models/transformers/monkey_patch.py (hunk 3 above)
Comment (PeterSH6) [reply verl/models/transformers/monkey_patch.py:46]: Let me check the lower bound of the transformer version

File: verl/workers/sharding_manager/fsdp_vllm.py [hunk 6]
```diff
@@ -66,6 +80,11 @@ def __enter__(self):
         # torch.cuda.empty_cache()
         # if torch.distributed.get_rank() == 0:
         # print(f'after model to cpu in sharding manager memory allocated: {torch.cuda.memory_allocated() / 1e9}GB, reserved: {torch.cuda.memory_reserved() / 1e9}GB')
+        
+        # important: need to manually set the random states of each tp to be identical. Otherwise, xperf_gpt will hang
```
Comment (vermouth1992) [inline verl/workers/sharding_manager/fsdp_vllm.py]: change comment from xperf_gpt to vllm

**Comment by PeterSH6** [comment]
Almost finished.

I wonder what kind of examples shall we add? We can add some scripts in the next PR.

File: verl/models/transformers/monkey_patch.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    return success_apply_monkey_patch
+
+
+from functools import lru_cache
+from packaging import version
+import importlib.metadata
+
+
+@lru_cache()
+def is_transformers_version_in_range(min_version: str, max_version: str) -> bool:
```
Comment (vermouth1992) [inline verl/models/transformers/monkey_patch.py:66]: this function can generalize to any lib

**Comment by xingyaoww** [comment]
Yes, it is able to produce; when enabling Ulysses context parallelism, set gradient_checkpointing_enable to False, and everything works. And turn it on will result in the above indexing error.

**Comment by PeterSH6** [comment]
@xingyaoww Cool! So you implemented Ulysses in the SFT trainer?

**Comment by xingyaoww** [comment]
@PeterSH6 yep! most changes are here: https://github.com/xingyaoww/verl/commits/dev

though I'm still testing it :) but so far it seems to work pretty well

**Comment by PeterSH6** [comment]
@xingyaoww It would be really nice.

I've seen your LoRA PR. It looks great.
It would be even better if you could create a PR for Ulysses + rmpad in SFTTrainer.
We really appreciate your effort!

**Comment by xingyaoww** [comment]
@PeterSH6 definitely -- draft PR up in https://github.com/volcengine/verl/pull/132

Will clean up the code there when LoRA is merged :)
